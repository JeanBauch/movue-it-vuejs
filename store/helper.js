import { createLocalVue } from '@vue/test-utils';
import Vuex from 'vuex';

import * as Challenges from '~/store/Challenges';
import ChallengesGetters from '~/store/Challenges/getters';
import ChallengesMutations from '~/store/Challenges/mutations';

import * as Countdown from '~/store/Countdown';

export const setupLocalVueStore = () => {
	const localVue = createLocalVue();
	localVue.use(Vuex);

	const setupModule = () => ({
		modules: {
			Countdown: {
				namespaced: true,
				state: Countdown.state(),
				getters: Countdown.getters,
				mutations: Countdown.mutations,
			},
			Challenges: {
				namespaced: true,
				state: {
					...Challenges.state(),
					allChallenges: [{
						type: 'type',
						description: 'description',
						amount: 10,
					}],
				},
				getters: ChallengesGetters,
				mutations: ChallengesMutations,
			},
		},
	});

	const configureStore = (countdownState = {}, challengesState = {}) => {
		const storeConfig = setupModule();
		storeConfig.modules.Countdown.state = {
			...storeConfig.modules.Countdown.state,
			...countdownState,
		};

		storeConfig.modules.Challenges.state = {
			...storeConfig.modules.Challenges.state,
			...challengesState,
		};

		const store = new Vuex.Store(storeConfig);
		return store;
	};

	return {
		localVue,
		configureStore,
	};
};
