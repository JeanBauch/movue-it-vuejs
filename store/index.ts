import { Mutations } from './Challenges/types';

export const actions = {
	nuxtServerInit ({ commit }: any, { app }: any) {
		const cookie = app.$cookiz.get('movueit');

		if (cookie) {
			commit(`Challenges/${Mutations.SAVE_COOOKIE_DATA}`, cookie);
		};
	},
};
