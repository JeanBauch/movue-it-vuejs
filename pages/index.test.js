jest.mock('~/utils', () => ({
	scrollToElement: jest.fn(),
	getRandomNumber: jest.fn(),
	playAudio: jest.fn(),
	sendNotification: jest.fn(),
}));

import { mount } from '@vue/test-utils';
import VueMeta from 'vue-meta';

import { setupLocalVueStore } from '~/store/helper';
import {
	scrollToElement,
	getRandomNumber,
	playAudio,
	sendNotification,
} from '~/utils';

import Countdown from '~/components/molecules/Countdown.vue';
import Index from './index.vue';

describe('Pages:index', () => {
	const { localVue, configureStore } = setupLocalVueStore();
	localVue.use(VueMeta, { keyName: 'head' });
	let mountConfig;

	beforeEach(() => {
		scrollToElement.mockClear();
		getRandomNumber.mockClear();
		playAudio.mockClear();
		sendNotification.mockClear();
		global.Notification = {
			requestPermission: jest.fn(),
			permission: 'default',
		};

		mountConfig = {
			localVue,
			stubs: {
				Profile: true,
				CompletedChallenges: true,
				Countdown: true,
				Card: true,
			},
		};
	});

	const buildWrapper = (countdownState = {}, challengesState = {}) => {
		const store = configureStore(countdownState, challengesState);
		return {
			...mountConfig,
			store,
		};
	};

	describe('Snapshots', () => {
		it('should render all child components and start a cycle button', () => {
			const config = buildWrapper();
			const wrapper = mount(Index, config);
			expect(wrapper.html()).toMatchSnapshot();
		});

		it('should render all child components and abandon cycle button', () => {
			const config = buildWrapper({ isActive: true });
			const wrapper = mount(Index, config);
			expect(wrapper.html()).toMatchSnapshot();
		});

		it('should render all child components and cycle completed button', () => {
			const config = buildWrapper({ hasCompleted: true });
			const wrapper = mount(Index, config);
			expect(wrapper.html()).toMatchSnapshot();
		});
	});

	describe('Meta info', () => {
		it('should have a meta title', () => {
			const config = buildWrapper();
			const wrapper = mount(Index, config);

			expect(wrapper.vm.$metaInfo.title).toBe('Home | movue.it');
		});
	});

	describe('Mounted', () => {
		it('should request Notification permissions when mounted', () => {
			global.Notification = {
				requestPermission: jest.fn(),
			};

			const config = buildWrapper();
			mount(Index, config);

			expect(global.Notification.requestPermission).toHaveBeenCalled();
		});

		it('should not request permissions if Notification doesnt exist', () => {
			delete global.Notification;

			const config = buildWrapper();
			mount(Index, config);

			expect(global.Notification).toBe(undefined);
		});
	});

	describe('Buttons clicks', () => {
		it('should start a cycle when start a cycle button is clicked', async () => {
			const config = buildWrapper();
			const wrapper = mount(Index, config);

			const button = wrapper.find('button');
			await button.trigger('click');

			const newButton = wrapper.find('button');
			expect(newButton.text()).toBe('Abandon cycle');
		});

		it('should abandon cycle when abandon cycle button is clicked', async () => {
			const config = buildWrapper({ isActive: true });
			const wrapper = mount(Index, config);

			const button = wrapper.find('button');
			await button.trigger('click');

			const newButton = wrapper.find('button');
			expect(newButton.text()).toBe('Start a cycle');
		});
	});

	describe('Emitted events', () => {
		it('should run getNewChallenge and play audio and send notification', async () => {
			global.Notification = {
				requestPermission: jest.fn(),
				permission: 'granted',
			};

			getRandomNumber.mockImplementationOnce(() => 1);

			const config = buildWrapper();
			const wrapper = mount(Index, config);

			const setCurrentChallengeIndex = jest.spyOn(wrapper.vm, 'setCurrentChallengeIndex');

			const countdown = wrapper.findComponent(Countdown);
			await countdown.vm.$emit('completed');

			const button = wrapper.find('button');

			expect(button.text()).toBe('Cycle completed');
			expect(setCurrentChallengeIndex).toHaveBeenCalledWith(1);
			expect(playAudio).toHaveBeenCalledWith('/notification.mp3');
			expect(sendNotification).toHaveBeenCalledWith('New Challenge!', {
				body: 'A new challenge has started! Go complete it!',
				icon: '/favicon.png',
			});
			expect(scrollToElement).toHaveBeenCalledWith('#challenge');
		});

		it('should run getNewChallenge but doesnt play audio and doesnt send notification', async () => {
			getRandomNumber.mockImplementationOnce(() => 1);

			const config = buildWrapper();
			const wrapper = mount(Index, config);

			const setCurrentChallengeIndex = jest.spyOn(wrapper.vm, 'setCurrentChallengeIndex');

			const countdown = wrapper.findComponent(Countdown);
			await countdown.vm.$emit('completed');

			const button = wrapper.find('button');

			expect(button.text()).toBe('Cycle completed');
			expect(setCurrentChallengeIndex).toHaveBeenCalledWith(1);
			expect(playAudio).not.toHaveBeenCalled();
			expect(sendNotification).not.toHaveBeenCalled();
			expect(scrollToElement).toHaveBeenCalledWith('#challenge');
		});
	});
});
