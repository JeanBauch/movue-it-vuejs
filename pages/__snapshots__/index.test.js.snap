// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages:index Snapshots should render all child components and abandon cycle button 1`] = `
"<div class=\\"w-full bg-background\\">
  <div class=\\"min-h-screen max-w-5xl mx-auto flex flex-col px-5 py-8 md:py-10\\">
    <header class=\\"pb-10 flex items-center w-full\\">
      <p class=\\"text-sm\\">
        0 xp
      </p>
      <div class=\\"mx-4 lg:mx-6 flex-1 bg-gray-line h-1 rounded-md\\">
        <div class=\\"relative bg-green h-1\\" style=\\"width: 0%;\\">
          <p class=\\"text-sm absolute -right-4 w-max top-2\\">
            0 xp (0%)
          </p>
        </div>
      </div>
      <p class=\\"text-sm\\">
        64 xp
      </p>
    </header>
    <section class=\\"flex flex-col lg:flex-row flex-1 lg:flex-none lg:mt-16 sm:gap-x-10 md:gap-x-20\\">
      <div class=\\"flex flex-col w-full lg:w-1/2\\">
        <profile-stub></profile-stub>
        <completedchallenges-stub></completedchallenges-stub>
        <countdown-stub></countdown-stub> <button class=\\"button abandon\\">
          Abandon cycle
        </button>
      </div>
      <card-stub id=\\"challenge\\" class=\\"w-full lg:w-1/2\\"></card-stub>
    </section>
  </div>
</div>"
`;

exports[`Pages:index Snapshots should render all child components and start a cycle button 1`] = `
"<div class=\\"w-full bg-background\\">
  <div class=\\"min-h-screen max-w-5xl mx-auto flex flex-col px-5 py-8 md:py-10\\">
    <header class=\\"pb-10 flex items-center w-full\\">
      <p class=\\"text-sm\\">
        0 xp
      </p>
      <div class=\\"mx-4 lg:mx-6 flex-1 bg-gray-line h-1 rounded-md\\">
        <div class=\\"relative bg-green h-1\\" style=\\"width: 0%;\\">
          <p class=\\"text-sm absolute -right-4 w-max top-2\\">
            0 xp (0%)
          </p>
        </div>
      </div>
      <p class=\\"text-sm\\">
        64 xp
      </p>
    </header>
    <section class=\\"flex flex-col lg:flex-row flex-1 lg:flex-none lg:mt-16 sm:gap-x-10 md:gap-x-20\\">
      <div class=\\"flex flex-col w-full lg:w-1/2\\">
        <profile-stub></profile-stub>
        <completedchallenges-stub></completedchallenges-stub>
        <countdown-stub></countdown-stub> <button class=\\"button start\\">
          Start a cycle
        </button>
      </div>
      <card-stub id=\\"challenge\\" class=\\"w-full lg:w-1/2\\"></card-stub>
    </section>
  </div>
</div>"
`;
